core/resources.rs

```diff
@@ -8,16 +8,23 @@
 
 use crate::error::bad_resource_id;
 use crate::error::AnyError;
+use crate::ZeroCopyBuf;
+use futures::Future;
 use std::any::type_name;
 use std::any::Any;
 use std::any::TypeId;
 use std::borrow::Cow;
 use std::collections::BTreeMap;
 use std::iter::Iterator;
+use std::pin::Pin;
 use std::rc::Rc;
 
+/// Returned by resource read/write/shutdown methods
+pub type AsyncResult<T> = Pin<Box<dyn Future<Output = Result<T, AnyError>>>>;
+
 /// All objects that can be store in the resource table should implement the
 /// `Resource` trait.
+/// TODO(@AaronO): investigate avoiding alloc on read/write/shutdown
```

Comment:
I am pretty certain that boxing the returned trait is unavoidable with trait methods.
It'd be nice though if we could avoid double boxing.